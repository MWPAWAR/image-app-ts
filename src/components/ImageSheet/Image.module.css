.image {
  border-radius: 10px;
  width: 100%;
  height: 230px;
}

.imageName {
  font-size: 14px;
  font-weight: 600;
  overflow: hidden;
  text-overflow: ellipsis;
  width: 80%;
}

.imageSize {
  font-size: 14px;
  color: #5f6368;
}

.imgNameAndFavtIconContainer {
  margin-top: 10px;
  display: flex;
}

.favtIcon {
  text-align: end;
  width: 20%;
}

.favtIconFilled:before {
  content: "♥";
  color: #5f6368;
  cursor: pointer;
  font-size: 18px;
}

.favtIconOutlined:before {
  content: "♡";
  color: #5f6368;
  cursor: pointer;
  font-size: 18px;
}
