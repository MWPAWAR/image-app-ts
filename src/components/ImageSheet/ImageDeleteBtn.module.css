.deleteBtnFooter {
  margin-top: 30px;
}
.button {
  background-color: #fff;
  width: 100%;
  padding: 10px;
  border: 1px solid #dadce0;
  cursor: pointer;
}