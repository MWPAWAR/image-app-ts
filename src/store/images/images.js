import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import getImages from './api';

export const fetchImages = createAsyncThunk('images/fetchImages', async () => {
  const response = await getImages();
  return response;
});

const imagesSlice = createSlice({
  name: 'images',
  initialState: { status: 'idle', images: [] },
  reducers: {
    toggleFavoriteImg(state, action) {
      const index = state.images.findIndex((image) => image.id === action.payload);
      if (typeof index !== 'undefined') {
        state.images[index].favorited = !state.images[index].favorited;
      }
    },
    deleteImg(state, action) {
      const index = state.images.findIndex((image) => image.id === action.payload);
      if (typeof index !== 'undefined') {
        state.images.splice(index, 1);
      }
    }
  },
  extraReducers(builder) {
    builder
      .addCase(fetchImages.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(fetchImages.fulfilled, (state, action) => {
        state.status = 'success';
        state.images = action.payload;
      })
      .addCase(fetchImages.rejected, (state) => {
        state.error = 'error';
      })
  }
});

export const { toggleFavoriteImg, deleteImg } = imagesSlice.actions;
export default imagesSlice.reducer;
