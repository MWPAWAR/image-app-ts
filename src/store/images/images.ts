import { createSlice, createAsyncThunk } from '@reduxjs/toolkit'
import { ImageListTypes } from '../../components/types'
import getImages from './api'
import { API_FETCH_STATUS } from '../../constants'
import {
  ImagesState,
  ImagesExtraReducerStateAction,
  ImagesReducerStateAction,
  ImagesFetchApiResponse,
} from './types'

export const fetchImages = createAsyncThunk<ImagesFetchApiResponse>(
  'images/fetchImages',
  async () => {
    const response: ImageListTypes = await getImages()
    return response
  }
)

const initialState: ImagesState = {
  status: API_FETCH_STATUS.IDLE,
  images: [],
}

const imagesSlice = createSlice({
  name: 'images',
  initialState,
  reducers: {
    toggleFavoriteImg(
      state: ImagesState,
      action: ImagesReducerStateAction
    ): void {
      const index: number | undefined = state.images.findIndex(
        (image) => image.id === action.payload.id
      )
      if (typeof index !== 'undefined') {
        state.images[index].favorited = !state.images[index].favorited
      }
    },
    deleteImg(state: ImagesState, action: ImagesReducerStateAction): void {
      const index: number | undefined = state.images.findIndex(
        (image) => image.id === action.payload.id
      )
      if (typeof index !== 'undefined') {
        state.images.splice(index, 1)
      }
    },
  },
  extraReducers(builder) {
    builder
      .addCase(fetchImages.pending, (state: ImagesState) => {
        state.status = API_FETCH_STATUS.LOADING
      })
      .addCase(
        fetchImages.fulfilled,
        (state: ImagesState, action: ImagesExtraReducerStateAction) => {
          state.status = API_FETCH_STATUS.SUCCESS
          state.images = action.payload
        }
      )
      .addCase(fetchImages.rejected, (state: ImagesState) => {
        state.status = API_FETCH_STATUS.ERROR
      })
  },
})

export const { toggleFavoriteImg, deleteImg } = imagesSlice.actions
export default imagesSlice.reducer
