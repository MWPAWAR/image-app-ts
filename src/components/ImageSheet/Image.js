import { useDispatch } from 'react-redux';
import imageStyles from './Image.module.css';
import classNames from 'classnames';
import { getMbsTextByBytes } from '../../utils/get-mbs-by-bytes';
import { toggleFavoriteImg } from '../../store/images/images';
import { resetSelectedImg } from '../../store/imageUiState/imageUiState';

const Image = ({ activeTab, selectedImage }) => {
  const dispatch = useDispatch();
  const handleFavtClick = () => {
    dispatch(toggleFavoriteImg(selectedImage.id));
    // This is to reset the selected img if the active tab is favorites.
    if (activeTab === 'favorites') dispatch(resetSelectedImg());
  };

  return (
    <div>
      <img
        src={selectedImage.url}
        className={imageStyles.image}
        alt={selectedImage.filename}
      />
      <div className={imageStyles.imgNameAndFavtIconContainer}>
        <div className={imageStyles.imageName}>{selectedImage.filename}</div>
        <div className={classNames(
          imageStyles.favtIcon,
          selectedImage.favorited ? imageStyles.favtIconFilled : imageStyles.favtIconOutlined
        )} onClick={handleFavtClick} />
      </div>
      <div className={imageStyles.imageSize}>{getMbsTextByBytes(selectedImage.sizeInBytes)}</div>
    </div>
  );
};

export default Image;
