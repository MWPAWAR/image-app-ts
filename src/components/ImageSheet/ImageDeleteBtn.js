import { useDispatch } from 'react-redux';
import imageDeleteBtnStyles from './ImageDeleteBtn.module.css';
import { deleteImg } from '../../store/images/images';
import { resetSelectedImg } from '../../store/imageUiState/imageUiState';

const ImageDeleteBtn = ({ selectedImage }) => {
  const dispatch = useDispatch();
  const handleDeleteClick = () => {
    dispatch(deleteImg(selectedImage.id));
    dispatch(resetSelectedImg(selectedImage.id));
  };

  return (
    <div className={imageDeleteBtnStyles.deleteBtnFooter}>
      <button
        className={imageDeleteBtnStyles.button}
        onClick={handleDeleteClick}
        >Delete</button>
    </div>
  )
};

export default ImageDeleteBtn;
