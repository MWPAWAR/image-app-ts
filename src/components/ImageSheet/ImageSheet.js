import { useSelector } from 'react-redux';
import { selectImageById } from '../../store/imageUiState/selectors';
import Image from './Image';
import ImageDeleteBtn from './ImageDeleteBtn';
import ImageDescription from './ImageDescription';
import ImageDetails from './ImageDetails';
import imageSheetStyles from './ImageSheet.module.css';

const ImageSheet = ({ activeTab }) => {
  const selectedImage = useSelector(selectImageById);

  return (
    <div className={imageSheetStyles.imageSheet}>
      <Image selectedImage={selectedImage} activeTab={activeTab} />
      <ImageDetails selectedImage={selectedImage} />
      <ImageDescription selectedImage={selectedImage} />
      <ImageDeleteBtn selectedImage={selectedImage} />
    </div>
  );
};

export default ImageSheet;
