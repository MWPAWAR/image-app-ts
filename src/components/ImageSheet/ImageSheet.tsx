import { useAppSelector } from '../../customHooks/hooks'
import { selectImageById } from '../../store/imageUiState/selectors'
import Image from './Image/Image'
import ImageDeleteBtn from './ImageDeleteBtn/ImageDeleteBtn'
import ImageDescription from './ImageDescription/ImageDescription'
import ImageDetails from './ImageDetails/ImageDetails'
import imageSheetStyles from './ImageSheet.module.css'
import { ImageSheetProps } from './types'

const ImageSheet = ({ activeTab }: ImageSheetProps) => {
  const selectedImage = useAppSelector(selectImageById)

  if (!selectedImage) return null

  return (
    <div className={imageSheetStyles.imageSheet}>
      <Image selectedImage={selectedImage} activeTab={activeTab} />
      <ImageDetails selectedImage={selectedImage} />
      <ImageDescription selectedImage={selectedImage} />
      <ImageDeleteBtn selectedImage={selectedImage} />
    </div>
  )
}

export default ImageSheet
