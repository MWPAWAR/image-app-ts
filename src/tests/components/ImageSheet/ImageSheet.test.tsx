import { MemoryRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer'
import configureStore from 'redux-mock-store'
import ImageSheet from '../../../components/ImageSheet/ImageSheet'
import { imageMock1 } from '../../mocks/image'
import { images } from '../../mocks/images'
import { toggleDrawer } from '../../test-utils'

describe('ImageSheet', () => {
  it('ImageSheet matches snapshot with data', () => {
    const mockStore = configureStore([])
    const store = mockStore({
      images: {
        images,
      },
      imageUiState: {
        selectedImg: imageMock1.id,
      },
    })
    jest.spyOn(store, 'dispatch')

    const ImageSheetComponent = () => (
      <MemoryRouter>
        <Provider store={store}>
          <ImageSheet activeTab="recentlyAdded" toggleDrawer={toggleDrawer} />
        </Provider>
      </MemoryRouter>
    )

    const tree = renderer.create(<ImageSheetComponent />)
    expect(tree).toMatchSnapshot()
  })

  it('ImageSheet matches snapshot with data', () => {
    const mockStore = configureStore([])
    const store = mockStore({
      images: {
        images,
      },
      imageUiState: {
        selectedImg: '',
      },
    })
    jest.spyOn(store, 'dispatch')

    const ImageSheetComponent = () => (
      <MemoryRouter>
        <Provider store={store}>
          <ImageSheet activeTab="recentlyAdded" toggleDrawer={toggleDrawer} />
        </Provider>
      </MemoryRouter>
    )

    const tree = renderer.create(<ImageSheetComponent />)
    expect(tree).toMatchSnapshot()
  })
})
