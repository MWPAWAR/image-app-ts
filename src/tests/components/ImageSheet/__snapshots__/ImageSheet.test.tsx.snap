// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageSheet ImageSheet matches snapshot with data 1`] = `
<div
  className="imageSheet"
>
  <div>
    <img
      alt="tennessee_female_rubber.jpg"
      className="image"
      src="https://agencyanalytics-api.vercel.app/images/0.jpg"
    />
    <div
      className="imgNameAndFavtIconContainer"
    >
      <div
        className="imageName"
      >
        tennessee_female_rubber.jpg
      </div>
      <div
        className="favtIcon favtIconFilled"
        data-testid="fav-icon"
        onClick={[Function]}
      />
    </div>
    <div
      className="imageSize"
    >
      4.6 MB
    </div>
  </div>
  <div
    className="imgDetails"
  >
    <div
      className="informationLabel"
    >
      Information
    </div>
    <div
      className="row"
    >
      <div
        className="label"
      >
        Uploaded by
      </div>
      <div
        className="value"
      >
        Ms. Jimmie Cole
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="label"
      >
        Created
      </div>
      <div
        className="value"
      >
        July 15, 2017
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="label"
      >
        Last modified
      </div>
      <div
        className="value"
      >
        December 16, 2022
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="label"
      >
        Dimensions
      </div>
      <div
        className="value"
      >
        4800 X 3200
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="label"
      >
        Resolutions
      </div>
      <div
        className="value"
      >
        72 X 72
      </div>
    </div>
  </div>
  <div
    className="imgDescription"
  >
    <div
      className="label"
    >
      Description
    </div>
    <div
      className="description"
    >
      Laboriosam eligendi inventore officia nemo. Quisquam explicabo voluptatem. Illo laborum facilis.
    </div>
  </div>
  <div
    className="deleteBtnFooter"
  >
    <button
      className="button"
      onClick={[Function]}
    >
      Delete
    </button>
  </div>
</div>
`;

exports[`ImageSheet ImageSheet matches snapshot with data 2`] = `null`;
